### PR TITLE
Track the amount of read data per row

### PR DESCRIPTION
**Pull request by jacek-lewandowski**
If an sstable is corrupted in a nasty way, we may read invalid cell sizes and try to read much more data for a row than we should. In rare scenarios this can lead even to OOMs.

This simple fix adds tracking and limiting the amount of data that is read per row. Row has its size stored in preamble which can be used as a limit. If the deserialization code tries to read more than that, it will simply fail with EOF which will prevent more serious problems later.

Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java [hunk 1]
```diff
@@ -49,53 +66,61 @@ public void reset(long count)
 
     public boolean readBoolean() throws IOException
     {
+        checkCanRead(1);
```
Comment (bereng) [inline src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: Do you want to use `TypeSizes` instead here?

File: test/unit/org/apache/cassandra/db/rows/UnfilteredSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            out.write(replacement);
+            out.close();
+            return out.asNewBuffer();
+        }
+        catch (IOException ex)
+        {
+            throw new RuntimeException(ex);
+        }
+    }
+}
```
Comment (bereng) [inline test/unit/org/apache/cassandra/db/rows/UnfilteredSerializerTest.java]: Missing new line?

**Comment by bereng** [comment]
We need CI runs and will you port this to 4.0 and 4.1 also?

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java (hunk 1 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: well, I don't think so, that would require also refactoring that in `bytesRead` updates

**Comment by jacek-lewandowski** [comment]
@bereng thanks for super quick response. I'll provide patches for older branches but probably it will be next week

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java (hunk 1 above)
Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: I think it is more clear to use the size number 1/2/4/8 ,which make the code more readable as the code below just add these numbers

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java [hunk 3]
```diff
@@ -129,22 +156,34 @@ public String readUTF() throws IOException
...
-        int skipped = source.skipBytes(n);
+        int skipped = source.skipBytes(limit < 0 ? n : (int) Math.min(limit - bytesRead, n));
         bytesRead += skipped;
         return skipped;
     }
+
+    @Inline
+    private void checkCanRead(int size) throws IOException
+    {
+        if (limit >= 0 && bytesRead + size > limit)
```
Comment (Maxwell-Guo) [inline src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: will limit just be zero？

File: test/unit/org/apache/cassandra/db/rows/UnfilteredSerializerTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+             DataInputBuffer replIn = new DataInputBuffer(replacement, false))
+        {
+            out.writeByte(origIn.readByte()); // flag
+            out.writeUnsignedVInt(origIn.readUnsignedVInt()); // row size
+            replIn.readByte(); // skip flag
+            replIn.readUnsignedVInt(); // skip row size
+            out.write(replacement);
+            out.close();
+            return out.asNewBuffer();
+        }
```
Comment (Maxwell-Guo) [inline test/unit/org/apache/cassandra/db/rows/UnfilteredSerializerTest.java]: I think a test for some corner case is needed such as what I said when limit is 0？

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java (hunk 3 above)
Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: I just test some times,it seems limit can not be 0 ~~~

Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: Yes, we always encode at least a flag

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java (hunk 1 above)
Comment (bereng) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: Not following. Why would you need to refactor anything? You guys think `1` is more readable than `TypeSizes.BOOL_SIZE` and on top of that you're hardcoding a value? It's like magic numbers where we have a proper abstraction ready unless I am missing sthg.

`checkCanRead(TypeSizes.BOOL_SIZE)` reads good in my eyes :shrug:

**Comment by jacek-lewandowski** [comment]
@bereng is it ok now?

File: src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java (hunk 1 above)
Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: I agree with your point of view. From the perspective of code readability, TypeSizes.BOOL_SIZE is definitely much better than hard code 1.
What I said above means that if the existing code is already written as 1/4/8, our newly added code also fills in 1/4/8, then it just echoes back and forth, so the readability is not bad.
Of course, if Jacek is willing to change the existing numbers to TypeSizes, I definitely support your point of view.

Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java]: (1)`checkCanRead(TypeSizes.BOOL_SIZE);
        byte b = source.readByte();
        bytesRead += TypeSizes.BOOL_SIZE;
        return b;`
 (2) `checkCanRead(1);
        byte b = source.readByte();
        bytesRead += 1;
        return b;`
(3)`checkCanRead(TypeSizes.BOOL_SIZE);
        byte b = source.readByte();
        bytesRead += 1;
        return b;`
I think the code itself should explain its behavior. 
just from my point of view, this order is (1) > (2) > (3) 
and it seems Jacek has already resolved this issue.

File: test/unit/org/apache/cassandra/db/rows/UnfilteredSerializerTest.java (hunk 4 above)
Comment (Maxwell-Guo) [reply test/unit/org/apache/cassandra/db/rows/UnfilteredSerializerTest.java]: No need for this

**Review by Maxwell-Guo — APPROVED** [review]

**Comment by bereng** [comment]
Approach looks good to me. CI and PRs for the rest of the versions are missing iiuc.

**Comment by jacek-lewandowski** [comment]
@bereng there was no point in running CI before code review, same for other PRs

**Comment by bereng** [comment]
Ha you're brave. I always run CI before...

**Comment by jacek-lewandowski** [comment]
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/720/workflows/7495bccf-c7f8-47db-a145-e0e5f6e508c2

**Comment by bereng** [comment]
^That is j11 only, j8 seems to not have been triggered?

**Comment by jacek-lewandowski** [comment]
https://app.circleci.com/pipelines/github/jacek-lewandowski/cassandra/720/workflows/f6fd1707-5cd1-4680-b234-0a79939fb4d2
